// @igniteui/angular-schematics/src/component/index_spec.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const schematics_1 = require("@angular-devkit/schematics");
const testing_1 = require("@angular-devkit/schematics/testing");
const cli_core_1 = require("@igniteui/cli-core");
const path = __importStar(require("path"));
const SchematicsTemplateManager_1 = require("../SchematicsTemplateManager");
const collectionPath = path.join(__dirname, "../collection.json");
describe("component", () => {
    beforeAll(() => {
        spyOn(cli_core_1.GoogleAnalytics, "post");
    });
    it("works", () => {
        const runner = new testing_1.SchematicTestRunner("schematics", collectionPath);
        const mockInst = {
            generateConfig: jasmine.createSpy(),
            packages: [],
            registerInProject: jasmine.createSpy(),
            templatePaths: []
        };
        const mockBaseTemplate = {
            id: "mock-template-id",
            name: "mock-template",
            description: "A mock template",
            delimiters: {
                content: { start: "{{", end: "}}" },
                path: { start: "[[", end: "]]" }
            },
            dependencies: ["mock-dependency"],
            framework: "angular",
            projectType: "ts",
            hasExtraConfiguration: true,
            templatePaths: ["/path/to/template"],
            generateConfig: jasmine.createSpy().and.returnValue({}),
            getExtraConfiguration: jasmine.createSpy().and.returnValue([]),
            setExtraConfiguration: jasmine.createSpy()
        };
        const mockProjectTemplate = Object.assign(Object.assign({}, mockBaseTemplate), { installModules: jasmine.createSpy().and.callFake(() => { }), upgradeIgniteUIPackages: jasmine.createSpy().and.returnValue(Promise.resolve(true)) });
        const mockTemplate = Object.assign(Object.assign({}, mockBaseTemplate), { components: ["mock-component"], controlGroup: "mock-group", listInComponentTemplates: true, listInCustomTemplates: true, packages: ["mock-package"], registerInProject: jasmine.createSpy() });
        const mockLib = {
            name: "mock-library",
            themes: ["mock-theme"],
            components: [{
                    name: "mock-component",
                    description: "A mock component",
                    group: "mock-group",
                    groupPriority: 1,
                    templates: [mockTemplate]
                }],
            projectIds: ["another-mock"],
            projects: [mockProjectTemplate],
            templates: [mockTemplate],
            projectType: "ts",
            generateTemplateFolderPath: "/path/to/templates",
            getCustomTemplateNames: jasmine.createSpy().and.returnValue([]),
            getTemplateByName: jasmine.createSpy().and.returnValue(mockTemplate),
            getTemplateById: jasmine.createSpy().and.returnValue(mockTemplate),
            getComponentByName: jasmine.createSpy().and.returnValue({
                name: "mock-component",
                description: "A mock component",
                group: "mock-group",
                groupPriority: 1,
                templates: [mockTemplate]
            }),
            getComponentGroupNames: jasmine.createSpy().and.returnValue(["mock-group"]),
            getComponentsByGroup: jasmine.createSpy().and.returnValue([{
                    name: "mock-component",
                    description: "A mock component",
                    group: "mock-group",
                    groupPriority: 1,
                    templates: [mockTemplate]
                }]),
            getComponentGroups: jasmine.createSpy().and.returnValue([{
                    name: "mock-group",
                    description: "A mock component group"
                }]),
            getCustomTemplates: jasmine.createSpy().and.returnValue([mockTemplate]),
            getProject: jasmine.createSpy().and.returnValue(mockProjectTemplate),
            hasProject: jasmine.createSpy().and.returnValue(false),
            hasTemplate: jasmine.createSpy().and.returnValue(false),
            registerTemplate: jasmine.createSpy()
        };
        const projLibSpy = spyOn(SchematicsTemplateManager_1.SchematicsTemplateManager.prototype, "getProjectLibrary");
        projLibSpy.and.returnValue(mockLib);
        const mockConfig = {
            project: {
                theme: "Custom"
            }
        };
        const projConfigSpy = spyOn(cli_core_1.ProjectConfig, "getConfig");
        projConfigSpy.and.returnValue(mockConfig);
        const tree = runner.runSchematic("component", { name: "my-combo", template: "combo", templateInst: mockInst, skipRoute: false }, schematics_1.Tree.empty());
        tree.then(state => {
            expect(mockInst.generateConfig).toHaveBeenCalledWith("my-combo", {});
            expect(mockInst.registerInProject).toHaveBeenCalledWith("", "my-combo", { skipRoute: false, modulePath: undefined });
            expect(projLibSpy).toHaveBeenCalledWith("angular", "igx-ts");
            expect(mockLib.hasTemplate).toHaveBeenCalledWith("combo");
            expect(mockLib.getTemplateById).toHaveBeenCalledWith("combo");
            expect(state.files).toEqual([]);
        });
    });
});
