// @igniteui/angular-templates/igx-ts/projects/_base_with_home/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseWithHomeProject = void 0;
const path = __importStar(require("path"));
const _base_1 = require("../_base");
class BaseWithHomeProject extends _base_1.BaseIgxProject {
    constructor() {
        super(...arguments);
        this.id = "base-with-home";
        this.name = "Base with home";
        this.description = "Empty project layout structure for Ignite UI for Angular with home page";
        this.dependencies = [];
        this.framework = "angular";
        this.projectType = "igx-ts";
        this.hasExtraConfiguration = false;
    }
    get templatePaths() {
        return [...super.templatePaths, path.join(__dirname, "files")];
    }
}
exports.BaseWithHomeProject = BaseWithHomeProject;
exports.default = new BaseWithHomeProject();
