// @igniteui/angular-templates/igx-ts/projects/_base/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseIgxProject = void 0;
const cli_core_1 = require("@igniteui/cli-core");
const path = __importStar(require("path"));
class BaseIgxProject {
    constructor() {
        this.id = "base";
        this.name = "base";
        this.description = "Empty project layout structure for Ignite UI for Angular";
        this.dependencies = [];
        this.framework = "angular";
        this.projectType = "igx-ts";
        this.hasExtraConfiguration = false;
        this.delimiters = {
            content: {
                end: `%>`,
                start: `<%=`
            },
            path: {
                end: `__`,
                start: `__`
            }
        };
        this.CUSTOM_THEME = `/* See: https://www.infragistics.com/products/ignite-ui-angular/angular/components/themes/sass/index */
@use "igniteui-angular/theming" as *;

$primary: #09f !default;
$secondary: #4db8ff !default;
$surface: #fff !default;

$app-palette: palette($primary, $secondary, $surface);

/* autoprefixer grid: on */

@include core();
@include typography($font-family: $material-typeface, $type-scale: $material-type-scale);
@include theme($app-palette);
`;
        this.DEFAULT_THEME = `,
              "node_modules/igniteui-angular/styles/igniteui-angular.css"`;
    }
    get templatePaths() {
        return [path.join(__dirname, "files")];
    }
    installModules() {
        throw new Error("Method not implemented.");
    }
    upgradeIgniteUIPackages(projectPath, packagePath) {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, cli_core_1.updateWorkspace)(projectPath);
        });
    }
    getExtraConfiguration() {
        return [];
    }
    setExtraConfiguration(extraConfigKeys) { }
    generateConfig(name, theme, ...options) {
        const config = {
            name,
            theme,
            cliVersion: cli_core_1.Util.version(),
            CustomTheme: "",
            dashName: cli_core_1.Util.lowerDashed(name),
            DefaultTheme: "",
            dot: ".",
            path: name,
            projectTemplate: this.id,
            igxPackage: cli_core_1.NPM_ANGULAR,
            yamlDefaultBranch: "<%=yaml-default-branch%>", // the placeholder will be evaluated by CodeGen
            ApplicationTitle: "<%=ApplicationTitle%>" // the placeholder will be evaluated by CodeGen
        };
        switch (theme) {
            case "Custom":
                config["CustomTheme"] = this.CUSTOM_THEME;
                config["themePath"] = "Custom";
                break;
            case "Default":
            default:
                config["DefaultTheme"] = this.DEFAULT_THEME;
                config["themePath"] = "node_modules/igniteui-angular/styles/igniteui-angular.css";
                break;
        }
        return config;
    }
    getVariablesConfig(name, theme) {
        return {
            name,
            theme,
            "cliVersion": cli_core_1.Util.version(),
            "dash-name": cli_core_1.Util.lowerDashed(name),
            "description": this.description,
            "dot": ".",
            "path": name,
            "projectTemplate": this.id,
            "yamlDefaultBranch": this.id === "base" ? "<%=yaml-default-branch%>" : "main"
        };
    }
}
exports.BaseIgxProject = BaseIgxProject;
