// @igniteui/cli-core/util/GlobalConstants.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.REACT_ROUTER_DOM_MODULE = exports.REACT_APP_ROUTING_NAME = exports.REACT_ROUTER_DOM_REDIRECT = exports.WC_ANCHOR_ELEMENT = exports.WC_APP_ROUTING_NAME = exports.ANCHOR_ELEMENT = exports.NG_CONFIG_TESTING_MODULE = exports.NG_HTTP_CLIENT_PROVIDER = exports.NG_COMMON_HTTP_PACKAGE = exports.NG_ROUTER_PACKAGE = exports.NG_DECORATOR_PROVIDERS = exports.PROVIDE_ROUTER = exports.WITH_COMPONENT_INPUT_BINDING = exports.NG_FOR_ROOT_IDENTIFIER_NAME = exports.NG_SA_DECORATOR_NAME = exports.NG_MODULE_DECORATOR_NAME = exports.TS_PRINTER_OPTIONS = exports.TRUE_CLAUSE = exports.UNDERSCORE_TOKEN = exports.SIDE_EFFECTS_IMPORT_TEMPLATE_NAME = exports.IMPORT_IDENTIFIER_NAME = exports.THEN_IDENTIFIER_NAME = exports.ROUTES_VARIABLE_NAME = void 0;
const ts = __importStar(require("typescript"));
const os_1 = require("os");
// TypeScript
exports.ROUTES_VARIABLE_NAME = 'routes';
exports.THEN_IDENTIFIER_NAME = 'then';
exports.IMPORT_IDENTIFIER_NAME = 'import';
exports.SIDE_EFFECTS_IMPORT_TEMPLATE_NAME = 'side_effects_import';
exports.UNDERSCORE_TOKEN = '_';
exports.TRUE_CLAUSE = 'true';
exports.TS_PRINTER_OPTIONS = {
    newLine: os_1.EOL === '\n'
        ? ts.NewLineKind.LineFeed
        : ts.NewLineKind.CarriageReturnLineFeed,
};
// Angular
exports.NG_MODULE_DECORATOR_NAME = 'NgModule';
exports.NG_SA_DECORATOR_NAME = 'Component';
exports.NG_FOR_ROOT_IDENTIFIER_NAME = 'forRoot';
exports.WITH_COMPONENT_INPUT_BINDING = 'withComponentInputBinding';
exports.PROVIDE_ROUTER = 'provideRouter';
exports.NG_DECORATOR_PROVIDERS = 'providers';
exports.NG_ROUTER_PACKAGE = '@angular/router';
exports.NG_COMMON_HTTP_PACKAGE = '@angular/common/http';
exports.NG_HTTP_CLIENT_PROVIDER = 'provideHttpClient';
exports.NG_CONFIG_TESTING_MODULE = 'configureTestingModule';
exports.ANCHOR_ELEMENT = {
    name: 'path',
    value: ts.factory.createStringLiteral('**'),
};
// WC
exports.WC_APP_ROUTING_NAME = 'app-routing.ts';
exports.WC_ANCHOR_ELEMENT = {
    name: 'path',
    value: ts.factory.createStringLiteral('(.*)'),
};
// React
exports.REACT_ROUTER_DOM_REDIRECT = 'redirect';
exports.REACT_APP_ROUTING_NAME = 'app-routes.tsx';
exports.REACT_ROUTER_DOM_MODULE = 'react-router-dom';
